File: typescript/tests/strip-governance.test.ts
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

/**
 * Strip-governance regression tests.
 *
 * TypeScript bundles governance unconditionally (consumers tree-shake unused
 * exports). These tests verify that:
 *
 * 1. Core telemetry features work without calling any governance APIs.
 * 2. Governance symbols live in isolated modules (classification, consent,
 *    receipts) that are only needed when explicitly imported.
 * 3. Core modules have no static import dependency on governance modules.
 */

import { beforeEach, describe, expect, it } from 'vitest';
import {
  counter,
  event,
  gauge,
  getHealthSnapshot,
  getLogger,
  getQueuePolicy,
  getSamplingPolicy,
  histogram,
  resetTelemetryState,
  setQueuePolicy,
  setSamplingPolicy,
  setupTelemetry,
  shutdownTelemetry,
} from '../src/index';

beforeEach(() => {
  resetTelemetryState();
});

describe('core telemetry without governance', () => {
  it('setup and shutdown succeed without governance', () => {
    // setupTelemetry returns void; we verify it does not throw.
    expect(() => setupTelemetry()).not.toThrow();
    expect(() => shutdownTelemetry()).not.toThrow();
  });

  it('getLogger returns a working logger', () => {
    const log = getLogger('strip-governance-test');
    expect(log).toBeDefined();
    // logger.info takes (obj, msg?) where obj is Record<string, unknown>
    log.info({ event: 'no_governance.test.logged', extra: 'value' });
  });

  it('counter instrument works without governance', () => {
    const c = counter('no_governance.test.counter');
    expect(c).toBeDefined();
    c.add(1);
  });

  it('gauge instrument works without governance', () => {
    const g = gauge('no_governance.test.gauge');
    expect(g).toBeDefined();
    g.set(42);
  });

  it('histogram instrument works without governance', () => {
    const h = histogram('no_governance.test.histogram');
    expect(h).toBeDefined();
    h.record(100);
  });

  it('health snapshot works without governance', () => {
    const snap = getHealthSnapshot();
    expect(snap.logsEmitted).toBeGreaterThanOrEqual(0);
    expect(snap.logsDropped).toBeGreaterThanOrEqual(0);
    expect(snap.tracesEmitted).toBeGreaterThanOrEqual(0);
    expect(snap.metricsEmitted).toBeGreaterThanOrEqual(0);
    expect(snap.setupError).toBeNull();
  });

  it('sampling policy roundtrip works without governance', () => {
    const policy = { defaultRate: 0.5, overrides: {} };
    setSamplingPolicy('logs', policy);
    const got = getSamplingPolicy('logs');
    expect(got.defaultRate).toBe(0.5);
    // Reset
    setSamplingPolicy('logs', { defaultRate: 1.0, overrides: {} });
  });

  it('queue policy roundtrip works without governance', () => {
    // QueuePolicy fields: maxLogs, maxTraces, maxMetrics
    setQueuePolicy({ maxLogs: 100, maxTraces: 100, maxMetrics: 100 });
    const got = getQueuePolicy();
    expect(got.maxLogs).toBe(100);
    // Reset
    setQueuePolicy({ maxLogs: 0, maxTraces: 0, maxMetrics: 0 });
  });

  it('event schema validation works without governance', () => {
    const e = event('auth', 'login', 'success');
    expect(e.event).toBe('auth.login.success');
    expect(e.domain).toBe('auth');
    expect(e.action).toBe('login');
    expect(e.status).toBe('success');
    expect(e.resource).toBeUndefined();
  });
});

describe('governance module isolation', () => {
  it('governance modules are in separate source files', async () => {
    // Governance must be in isolated modules, not bundled into core.
    // We verify this by importing them separately — if they were in the core
    // bundle, tree-shaking would not work.
    const { setConsentLevel, getConsentLevel } = await import('../src/consent');
    const { registerClassificationRules, getClassificationPolicy } =
      await import('../src/classification');
    const { enableReceipts } = await import('../src/receipts');

    expect(typeof setConsentLevel).toBe('function');
    expect(typeof getConsentLevel).toBe('function');
    expect(typeof registerClassificationRules).toBe('function');
    expect(typeof getClassificationPolicy).toBe('function');
    expect(typeof enableReceipts).toBe('function');
  });

  it('pii module does not statically require classification', async () => {
    // The PII sanitizer must work even when no classification rules are registered.
    const { sanitizePayload, replacePiiRules } = await import('../src/pii');

    replacePiiRules([]);
    const payload: Record<string, unknown> = {
      user: 'alice',
      password: 's3cr3t', // pragma: allowlist secret
      token: 'abc123',
    };
    sanitizePayload(payload);

    expect(payload['password']).toBe('***');
    expect(payload['token']).toBe('***');
    expect(payload['user']).toBe('alice');
    // No classification labels injected without registered rules
    expect(payload['__password__class']).toBeUndefined();
  });
});

File: go/strip_governance_test.go
// SPDX-FileCopyrightText: Copyright (C) 2026 provide.io llc
// SPDX-License-Identifier: Apache-2.0
// SPDX-Comment: Part of provide-telemetry.

//go:build nogovernance

// strip_governance_test.go — regression tests for the no-governance build.
// Run with: go test -tags nogovernance ./...
//
// These tests assert that core telemetry features work correctly when
// classification, consent, and receipts modules are excluded.
package telemetry

import (
	"context"
	"testing"
)

// TestNoGovernance_SetupShutdownRoundtrip verifies setup/shutdown works without governance.
func TestNoGovernance_SetupShutdownRoundtrip(t *testing.T) {
	ctx := context.Background()
	_ = ShutdownTelemetry(ctx)
	cfg, err := SetupTelemetry()
	if err != nil {
		t.Fatalf("setup must succeed without governance: %v", err)
	}
	if cfg.ServiceName == "" {
		t.Error("config must include a service name")
	}
	_ = ShutdownTelemetry(ctx)
}

// TestNoGovernance_LoggerUsable verifies the logger works without governance.
func TestNoGovernance_LoggerUsable(t *testing.T) {
	ctx := context.Background()
	log := GetLogger(ctx, "no-governance-test")
	if log == nil {
		t.Fatal("GetLogger must return a non-nil logger")
	}
	// No panic = success
	log.Info("no_governance.test.logged")
}

// TestNoGovernance_PIIRedactsDefaultSensitiveKeys verifies that default-sensitive
// keys are masked even without the classification module.
func TestNoGovernance_PIIRedactsDefaultSensitiveKeys(t *testing.T) {
	ReplacePIIRules(nil)
	payload := map[string]any{
		"user":     "alice",
		"password": "s3cr3t",
		"token":    "abc123",
	}
	result := SanitizePayload(payload, true, 3)
	if result["password"] != "***" {
		t.Errorf("password must be redacted, got %v", result["password"])
	}
	if result["token"] != "***" {
		t.Errorf("token must be redacted, got %v", result["token"])
	}
	if result["user"] != "alice" {
		t.Errorf("non-sensitive key must pass through, got %v", result["user"])
	}
	// No classification labels without governance
	if _, found := result["__password__class"]; found {
		t.Error("classification labels must be absent without governance")
	}
}

// TestNoGovernance_HealthSnapshotAvailable verifies health counters work without governance.
func TestNoGovernance_HealthSnapshotAvailable(t *testing.T) {
	snap := GetHealthSnapshot()
	if snap.LogsEmitted < 0 {
		t.Error("LogsEmitted must be non-negative")
	}
	if snap.LogsDropped < 0 {
		t.Error("LogsDropped must be non-negative")
	}
	if snap.TracesEmitted < 0 {
		t.Error("TracesEmitted must be non-negative")
	}
}

// TestNoGovernance_SamplingPolicyRoundtrip verifies sampling works without governance.
func TestNoGovernance_SamplingPolicyRoundtrip(t *testing.T) {
	policy := SamplingPolicy{DefaultRate: 0.5}
	if _, err := SetSamplingPolicy("logs", policy); err != nil {
		t.Fatalf("SetSamplingPolicy must not error: %v", err)
	}
	got, err := GetSamplingPolicy("logs")
	if err != nil {
		t.Fatalf("GetSamplingPolicy must not error: %v", err)
	}
	if got.DefaultRate != 0.5 {
		t.Errorf("expected default_rate=0.5, got %v", got.DefaultRate)
	}
	// Reset
	_, _ = SetSamplingPolicy("logs", SamplingPolicy{DefaultRate: 1.0})
}

// TestNoGovernance_CounterUsable verifies metric counters work without governance.
func TestNoGovernance_CounterUsable(t *testing.T) {
	ctx := context.Background()
	c := NewCounter("no_governance.test.counter")
	if c == nil {
		t.Fatal("NewCounter must return a non-nil instrument")
	}
	// No panic on increment
	c.Add(ctx, 1)
}
